pkg/cloudprovider/providers/aws/aws_loadbalancer.go

```diff
@@ -1289,6 +1237,75 @@ func (c *Cloud) ensureLoadBalancer(namespacedName types.NamespacedName, loadBala
 	return loadBalancer, nil
 }
 
+// syncElbListeners computes a plan to reconcile the desired vs actual state of the listeners on an ELB
+// NOTE: there exists an O(nlgn) implementation for this function. However, as the default limit of
+//       listeners per elb is 100, this implementation is reduced from O(m*n) => O(n).
+func syncElbListeners(loadBalancerName string, listeners []*elb.Listener, listenerDescriptions []*elb.ListenerDescription) (*elb.CreateLoadBalancerListenersInput, *elb.DeleteLoadBalancerListenersInput) {
+	foundSet := make(map[int]bool)
+	deleteRequest := &elb.DeleteLoadBalancerListenersInput{
+		LoadBalancerName: aws.String(loadBalancerName),
+	}
+	createRequest := &elb.CreateLoadBalancerListenersInput{
+		LoadBalancerName: aws.String(loadBalancerName),
+	}
+
+	for _, listenerDescription := range listenerDescriptions {
+		actual := listenerDescription.Listener
+		if actual == nil {
+			klog.Warning("Ignoring empty listener in AWS loadbalancer: ", loadBalancerName)
+			continue
+		}
+
+		for i, expected := range listeners {
+			if elbListenersAreEqual(actual, expected) {
+				// The current listener on the actual
+				// elb is in the set of desired listeners.
+				foundSet[i] = true
+				break
+			}
+			if i == len(listeners)-1 {
+				// As the current listener on the actual elb is not in
+				// the set of desired listeners, we need to remove it.
+				deleteRequest.LoadBalancerPorts = append(deleteRequest.LoadBalancerPorts, actual.LoadBalancerPort)
+			}
+		}
+	}
+
+	for i := range listeners {
+		if !foundSet[i] {
+			createRequest.Listeners = append(createRequest.Listeners, listeners[i])
+		}
+	}
+
+	return createRequest, deleteRequest
+}
+
+func elbListenersAreEqual(actual, expected *elb.Listener) bool {
+	if actual == nil && expected == nil {
```

Comment:
I see the value in this check, but given how complicated this method is already I worry about adding checks like this that would change the logic for ensureLoadBalancer unexpectedly.

Reply:
Would be a different story if we had coverage for the rest of the method

Reply:
I don't think this changes the logic. Rather, it saves a panic. If this check wasn't there (not that it's it's own method) and you pass in nil, it'll panic.

Reply:
That's fair, but doing the nil checks here is changing the behavior from returning true/false based on nil checks vs letting the controller crash and restart fresh. Can we determine with confidence which is the correct behavior at this point?

Reply:
> vs letting the controller crash and restart fresh
this would send the controller into an unrecoverable crash loop correct?
> but doing the nil checks here is changing the behavior 
isn't the desired behavior here to check to see if the listeners are equal. If so, then doing these nil checks are doing just that and enforcing this behavior even in edge cases.

I don't think it'd be okay to let the app go into a crash loop. We should either decide to (1) bubble up an error if it's nil (2) continue and ignore if it's nil. If we do (2), this method must take in non-pointer values.

Reply:
Alright so it looks like in the old implementation we do a `if actual == nil` check and so it should never be nil. For `expected`, I think it's safe to `continue`. However, since, this is in it's own method now, we need to handle the nil case IMO but given the nil checks in the calling function, it won't affect current behavior.

Reply:
> this would send the controller into an unrecoverable crash loop correct?

Yup, the controller would crash & restart but it may not necessarily end up in a crash _loop_. My point here was: let's try to reduce the number of things we are changing in this PR given the amount of test coverage here.  

Also worth mentioning that even with the nil checks, if both `expected` and `actual` are both `nil`, `elbListenersAreEqual` returns `true` so you would be adding a nil entry to a map and that would result in a panic anyways

Reply:
> let's try to reduce the number of things we are changing in this PR given the lack of test coverage here

Alright sounds good. If the goal is just to minimize the changes and just add tests, we can do that in this PR and worry about this edge case in another PR. I'll remove the nil checks :)